From fce3d2eb3a27ecf8f10d61a0175f2e7195e1388b Mon Sep 17 00:00:00 2001
From: Andy Duan <fugang.duan@nxp.com>
Date: Fri, 17 Aug 2018 17:07:06 +0800
Subject: [PATCH 4425/5242] MLK-19251-01 gpio: max732x: add output IO default
 voltage set

commit  7728ef704181a4fd00b08cb609daca07403d04cd from
https://source.codeaurora.org/external/imx/linux-imx.git

Add output IOs defalut voltage set in device tree by add property like:
	out-default = /bits/ 16 <mask val>;

Reviewed-by: Haibo Chen <haibo.chen@nxp.com>
Signed-off-by: Fugang Duan <fugang.duan@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../devicetree/bindings/gpio/gpio-max732x.txt      |    1 +
 drivers/gpio/gpio-max732x.c                        |   17 +++++++++++++++++
 2 files changed, 18 insertions(+)

diff --git a/Documentation/devicetree/bindings/gpio/gpio-max732x.txt b/Documentation/devicetree/bindings/gpio/gpio-max732x.txt
index 5fdc843..c79dcaa 100644
--- a/Documentation/devicetree/bindings/gpio/gpio-max732x.txt
+++ b/Documentation/devicetree/bindings/gpio/gpio-max732x.txt
@@ -32,6 +32,7 @@ Optional properties:
     - second cell is used to specify flags
   - interrupt-parent: phandle of the parent interrupt controller.
   - interrupts: Interrupt specifier for the controllers interrupt.
+  - out-default: set the output IO default voltage. Exp: out-default = /bits/ 16 <mask val>;
 
 Please refer to gpio.txt in this directory for details of the common GPIO
 bindings used by client devices.
diff --git a/drivers/gpio/gpio-max732x.c b/drivers/gpio/gpio-max732x.c
index 321dce6f..c184b09 100644
--- a/drivers/gpio/gpio-max732x.c
+++ b/drivers/gpio/gpio-max732x.c
@@ -81,6 +81,12 @@
 #define INT_CAPS(x)	(((uint64_t)(x)) << 32)
 
 enum {
+	OUTPUT_MASK,
+	OUTPUT_VAL,
+	OUTPUT_NUM,
+};
+
+enum {
 	MAX7319,
 	MAX7320,
 	MAX7321,
@@ -626,6 +632,8 @@ static int max732x_probe(struct i2c_client *client,
 	struct i2c_client *c;
 	uint16_t addr_a, addr_b;
 	int ret, nr_port;
+	u16 out_set[OUTPUT_NUM];
+	unsigned long mask, val;
 
 	pdata = dev_get_platdata(&client->dev);
 	node = client->dev.of_node;
@@ -711,6 +719,15 @@ static int max732x_probe(struct i2c_client *client,
 	}
 
 	i2c_set_clientdata(client, chip);
+
+	/* set the output IO default voltage */
+	if (!of_property_read_u16_array(node, "out-default", out_set,
+					ARRAY_SIZE(out_set))) {
+		mask = out_set[OUTPUT_MASK] & chip->dir_output;
+		val = out_set[OUTPUT_VAL];
+		max732x_gpio_set_multiple(&chip->gpio_chip, &mask, &val);
+	}
+
 	return 0;
 
 out_failed:
-- 
1.7.9.5

